lib/srv/db/cassandra/protocol/conn.go

```diff
@@ -0,0 +1,213 @@
+func (c *Conn) readFrame(r io.Reader) (*frame.Frame, error) {
+	fr, err := c.frameCodec.DecodeFrame(r)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	if startup, ok := fr.Body.Message.(*message.Startup); ok {
+		compression := startup.GetCompression()
+		c.frameCodec = frame.NewRawCodecWithCompression(client.NewBodyCompressor(compression))
+		c.segmentCode = segment.NewCodecWithCompression(client.NewPayloadCompressor(compression))
+		// If moderate framing layout is supported all received from a client after Startup message should
+		// use segment encoding.
+		c.modernLayoutRead = fr.Header.Version.SupportsModernFramingLayout()
+	}
+
+	return fr, nil
+}
+
+// readSegment is used to read segments from the connection.
+// If frame is not self-contained, segments are split into multiple frames.
+// Read the segments till received bodyLength bytes.
+func (c *Conn) readSegment(r io.Reader) (*frame.Frame, error) {
+	previousSegment := bytes.Buffer{}
+	expectedSegmentSize := 0
+
+	for {
+		seg, err := c.segmentCode.DecodeSegment(r)
+		if err != nil {
+			return nil, trace.Wrap(err)
+		}
+		if seg.Header.IsSelfContained {
+			fr, err := c.readFrame(bytes.NewReader(seg.Payload.UncompressedData))
+			if err != nil {
+				return nil, trace.Wrap(err)
+			}
+			return fr, nil
+		}
+		// Otherwise read the frame size and keep reading until we read all data.
+		// Segments are always delivered in order.
+		if expectedSegmentSize == 0 {
+			frameHeader, err := c.frameCodec.DecodeHeader(bytes.NewReader(seg.Payload.UncompressedData))
+			if err != nil {
+				return nil, trace.Wrap(err)
+			}
+			expectedSegmentSize = int(primitive.FrameHeaderLengthV3AndHigher + frameHeader.BodyLength)
+		}
+		// Append another segment
+		if _, err := previousSegment.Write(seg.Payload.UncompressedData); err != nil {
+			return nil, trace.Wrap(err)
+		}
+		// Return the frame after reading all segments.
+		if expectedSegmentSize == previousSegment.Len() {
+			fr, err := c.readFrame(&previousSegment)
+			if err != nil {
+				return nil, trace.Wrap(err)
+			}
+			return fr, nil
+		}
+	}
+}
+
+// writeFrameWithModernLayout is used to write frame to the connection.
+func (c *Conn) writeFrameWithModernLayout(outgoing *frame.Frame) error {
+	if c.modernLayoutWrite {
+		return trace.Wrap(c.writeSegment(outgoing, c.Conn))
+	}
+	return trace.Wrap(c.writeFrame(outgoing, c.Conn))
+}
+
+// writeFrame is used to write frame to the connection.
+func (c *Conn) writeFrame(outgoing *frame.Frame, wr io.Writer) error {
+	err := c.frameCodec.EncodeFrame(outgoing, wr)
+	return trace.Wrap(err)
+}
+
+// writeSegment is used to write segments to the connection.
+func (c *Conn) writeSegment(outgoing *frame.Frame, wr io.Writer) error {
+	var buff bytes.Buffer
+	if err := c.writeFrame(outgoing, &buff); err != nil {
+		return trace.Wrap(err)
+	}
+	seg := &segment.Segment{
+		Header:  &segment.Header{IsSelfContained: true},
+		Payload: &segment.Payload{UncompressedData: buff.Bytes()},
+	}
```

Comment:
Assuming in our use cases (mostly just for auth and error?) self contained is always enuf?

Reply:
Yes, The writeSegment function is used only during handling manual Cassandra handshake so it should be enough for handling the handshake. 
Also this function was borrow from:  https://github.com/datastax/go-cassandra-native-protocol/blob/5e8aad05cf90f30499e5873f70b37dbc175306fd/client/client.go#L370